### PR TITLE
[ADD][13.0] new module report_layout_config

### PR DESCRIPTION
**Pull request by Tonow-c2c**
image

------

### Result:
[Draft Invoice (_ 2)-5.pdf](https://github.com/OCA/reporting-engine/files/4046543/Draft.Invoice._.2.-5.pdf)

### CODE REVIEW

File: report_layout_config/views/res_company.xml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+        <record id="view_company_form" model="ir.ui.view">
+            <field name="name">view.company.form</field>
+            <field name="model">res.company</field>
+            <field name="inherit_id" ref="base.view_company_form"/>
+            <field name="arch" type="xml">
+               <xpath expr="//field[@name='favicon']" position="after">
+                  <field name="full_header_img" widget="image"/>
```
Comment (grindtildeath) [inline report_layout_config/views/res_company.xml]: IMO these fields should be made available in the General Settings wizard under the Business Documents section and displayed only if `external_layout_images` is selected. (means you have to declare related fields on the `res.config.settings`)

File: report_layout_config/models/res_company.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import fields, models
+
+class ResCompany(models.Model):
+    _inherit = "res.company"
+
+    full_header_img = fields.Binary(string="Full header image", help="This image will replace all header.")
+    full_footer_img = fields.Binary(string="Full footer image", help="This image will replace all footer.")
```
Comment (grindtildeath) [inline report_layout_config/models/res_company.py]: Add attachment=True on both fields.

**Review by grindtildeath — CHANGES_REQUESTED** [review]
Two things. Moreover I don't think reporting-engine is the right repository for such a module, but I'm not sure which one would be the best. I guess OCA/web would fit better since we have a dependency on the module..

File: report_layout_config/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+{
+    "name": "Report layout configuration",
+    "summary": "Add possibility to easely modify the global report layout",
```
Comment (jcoux) [inline report_layout_config/__manifest__.py]:
```suggestion
    "summary": "Add possibility to easily modify the global report layout",
```

File: report_layout_config/data/report_layout.xml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <data>
```
Comment (jcoux) [inline report_layout_config/data/report_layout.xml]: `data` element is useless.

File: report_layout_config/models/__init__.py [hunk 5]
```diff
@@ -0,0 +1,3 @@
+
```
Comment (jcoux) [inline report_layout_config/models/__init__.py]:
```suggestion
```

**Comment by Tonow-c2c** [comment]
@grindtildeath for the module place @gurneyalex suggest me to put on this repo :-)

File: report_layout_config/models/base_document_layout.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+    @api.depends('report_layout_id')
+    def _compute_need_images_layout(self):
+        img_lay_id = self.env.ref(
+            'report_layout_config.external_layout_images_template'
+        ).id
+        for wizard in self:
+            wizard.need_images_layout = False
+            if wizard.external_report_layout_id.id == img_lay_id:
+                wizard.need_images_layout = True
```
Comment (jcoux) [inline report_layout_config/models/base_document_layout.py]:
```
    @api.depends('report_layout_id')
    def _compute_need_images_layout(self):
        img_lay = self.env.ref(
            'report_layout_config.external_layout_images_template'
        )
        for wizard in self:
            wizard.need_images_layout = (
                wizard.external_report_layout_id == img_lay_id
            )
```

File: report_layout_config/models/base_document_layout.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+    @api.depends('report_layout_id', 'logo', 'font', 'primary_color', 'secondary_color', 'full_footer_img', 'full_header_img')
+    def _compute_preview(self):
+        img_lay_id = self.env.ref(
+            'report_layout_config.external_layout_images_template'
+        ).id
+        for wizard in self:
+            if not wizard.report_layout_id:
+                wizard.preview = False
+            elif wizard.external_report_layout_id.id != img_lay_id:
```
Comment (jcoux) [inline report_layout_config/models/base_document_layout.py]:
```suggestion
            elif wizard.external_report_layout_id != img_lay:
```
And change related lines.

File: report_layout_config/models/base_document_layout.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+                super()._compute_preview()
```
Comment (jcoux) [inline report_layout_config/models/base_document_layout.py]:
```suggestion
                super(BaseDocumentLayout, wizard)._compute_preview()
```
Or remove loop and and `self.ensure_one()`

File: report_layout_config/readme/DESCRIPTION.rst [hunk 9]
```diff
@@ -0,0 +1,2 @@
+This module provides new report template.
```
Comment (jcoux) [inline report_layout_config/readme/DESCRIPTION.rst]:
```suggestion
This module provides new report template
```

File: report_layout_config/readme/DESCRIPTION.rst [hunk 10]
```diff
@@ -0,0 +1,2 @@
...
+With possibility to add image to replace header and footer.
```
Comment (jcoux) [inline report_layout_config/readme/DESCRIPTION.rst]:
```suggestion
with possibility to add image to replace header and footer.
```

File: report_layout_config/templates/report_templates.xml [hunk 11]
```diff
@@ -0,0 +1,136 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <template id="external_layout_images_template">
+      <div class="header">
+          <div class="row">
+              <div class="col-3 mb4">
+                  <img t-if="company.full_header_img" t-att-src="image_data_uri(company.full_header_img)" style="max-height: 60px;" alt="Logo"/>
+              </div>
+              <div class="col-9 text-right" style="margin-top:22px;" t-field="company.report_header" name="moto"/>
```
Comment (jcoux) [inline report_layout_config/templates/report_templates.xml]:
```suggestion
              <div class="col-9 text-right" style="margin-top:22px;" t-field="company.report_header" name="report_header"/>
```

File: report_layout_config/templates/report_templates.xml [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+
+      <div class="article o_report_layout_standard"  t-att-data-oe-model="o and o._name" t-att-data-oe-id="o and o.id" t-att-data-oe-lang="o and o.env.context.get('lang')">
+          <t t-call="web.address_layout"/>
+          <t t-raw="0"/>
+      </div>
+
+      <div class="footer o_standard_footer">
+          <div class="text-center">
+              <img t-if="company.full_footer_img" t-att-src="image_data_uri(company.full_footer_img)" style="max-width:500px;"  alt="Footer logo"/>
+
```
Comment (jcoux) [inline report_layout_config/templates/report_templates.xml]:
```suggestion
```

File: report_layout_config/templates/report_templates.xml [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+              <div t-if="report_type == 'pdf'" class="text-muted">
+                  Page: <span class="page"/> / <span class="topage"/>
+              </div>
+          </div>
+      </div>
+    </template>
+
+
+    <!-- Same as odoo/src/odoo/addons/base/wizard/base_document_layout_views.xml -->
```
Comment (jcoux) [inline report_layout_config/templates/report_templates.xml:88]: Can be cool to indicate by inline comments which parts are modified.

File: report_layout_config/templates/report_templates.xml [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+                                        </td>
+                                    </t>
+                                </tr>
+                            </table>
+                        </div>
+                    </div>
+                </div>
+            </div>
+
+          </div>
```
Comment (jcoux) [inline report_layout_config/templates/report_templates.xml]: Strange indentation.

**Review by jcoux — CHANGES_REQUESTED** [review]

**Review by jcoux — APPROVED** [review]

File: report_layout_config/models/base_document_layout.py [hunk 15]
```diff
@@ -35,17 +35,12 @@ def _compute_need_images_layout(self):
...
-        if not self.report_layout_id:
-            self.preview = False
-        elif self.external_report_layout_id != img_lay_id:
+        if not self.need_images_layout or not self.report_layout_id:
             super()._compute_preview()
         else:
             ir_qweb = self.env["ir.qweb"]
             qweb_ctx = self.env["ir.ui.view"]._prepare_qcontext()
-            qweb_ctx.update({"company": self})
+            qweb_ctx.update({"company": self, "res_company": self.company_id})
```
Comment (grindtildeath) [inline report_layout_config/models/base_document_layout.py]: Looks like this change was unneeded.

**Review by grindtildeath — APPROVED** [review]
LG :+1:

**Pull request merged**